src/transformers/models/molmo/modular_molmo.py

```diff
@@ -0,0 +1,2289 @@
+        **kwargs,
+    ):
+        r"""
+        Instantiate a [`MolmoConfig`] (or a derived class) from molmo text model configuration, molmo vision model
+        configuration and molmo pooling module conffiguration.
+
+        Returns:
+            [`MolmoConfig`]: An instance of a configuration object
+        """
+
+        return cls(
+            text_config=text_config.to_dict(),
+            vision_config=vision_config.to_dict(),
+            pooling_config=pooling_config.to_dict(),
+            **kwargs,
+        )
+
+
+class MolmoCausalLMOutputWithPast(LlavaCausalLMOutputWithPast):
+    pass
+
+
+# swiglu activation
+class MolmoSwiGLU(nn.Module):
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x, gate = x.chunk(2, dim=-1)
+        return nn.functional.silu(gate) * x
+
+
+# text modules inherited from Qwen2
+class MolmoMLP(CLIPMLP):
+    def __init__(self, config):
+        super().__init__()
+        self.activation_fn = MolmoSwiGLU()
+        self.fc1 = nn.Linear(config.hidden_size, config.intermediate_size, bias=False)
+        self.fc2 = nn.Linear(config.intermediate_size // 2, config.hidden_size, bias=False)
+
+
+# We have different attention classes for the txt and the image components, they need to be propagated back correctly
+class MolmoTextAttention(Qwen2Attention):
+    pass
+
+
+class MolmoTextSdpaAttention(MolmoTextAttention, Qwen2SdpaAttention):
+    pass
+
+
+class MolmoTextFlashAttention2(MolmoTextAttention, Qwen2FlashAttention2):
+    pass
+
+
+MOLMO_TEXT_ATTENTION_CLASSES = {
+    "eager": MolmoTextAttention,
+    "sdpa": MolmoTextSdpaAttention,
+    "flash_attention_2": MolmoTextFlashAttention2,
+}
+
+
+class MolmoDecoderLayer(Qwen2DecoderLayer):
+    def __init__(self, config, layer_idx: int):
+        super().__init__()
+        self.mlp = MolmoMLP(config)
+        self.self_attn = MOLMO_TEXT_ATTENTION_CLASSES[config._attn_implementation](config, layer_idx)
+
+
+MOLMO_START_DOCSTRING = r"""
+    This model inherits from [`PreTrainedModel`]. Check the superclass documentation for the generic methods the
+    library implements for all its model (such as downloading or saving, resizing the input embeddings, pruning heads
+    etc.)
+
+    This model is also a PyTorch [torch.nn.Module](https://pytorch.org/docs/stable/nn.html#torch.nn.Module) subclass.
+    Use it as a regular PyTorch Module and refer to the PyTorch documentation for all matter related to general usage
+    and behavior.
+
+    Parameters:
+        config ([`MolmoConfig`]):
+            Model configuration class with all the parameters of the model. Initializing with a config file does not
+            load the weights associated with the model, only the configuration. Check out the
+            [`~PreTrainedModel.from_pretrained`] method to load the model weights.
+"""
+
+
+@add_start_docstrings(
+    "The bare Molmo Model outputting raw hidden-states without any specific head on top.",
+    MOLMO_START_DOCSTRING,
+)
```

Comment:
inheriting from almost any other pretrained model will allow you to remove this as well